### PR TITLE
UI - MultiSelect Input

### PR DESCRIPTION
**Pull request by nhaagen**
Use a list of labeled checkboxes for the user to take picks from a finite list.

bildschirmfoto_2018-08-10_11-46-13

More in the rules:
https://github.com/ILIAS-eLearning/ILIAS/pull/1136/files#diff-2d8571e3a451f51b51f9868081f6f7e8

### CODE REVIEW

File: src/UI/Component/Input/Field/Factory.php [hunk 1]
```diff
@@ -348,5 +347,41 @@ public function password($label, $byline = null);
...
+	 *      options where only one option can be selected. A Select MUST be used in this case
+	 *
+	 * ---
+	 * @param string 	$label
+	 * @param array<string,string> 	$options 	with the select options as value=>label.
+	 * @param string 	$byline
+	 *
+	 * @return \ILIAS\UI\Component\Input\Field\MultiSelect
+	 */
+	public function multiselect($label, array $options, $byline = null);
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Make two words of it by using lower camelCase: multSelect

File: src/UI/Component/Input/Field/Factory.php [hunk 2]
```diff
@@ -348,5 +347,41 @@ public function password($label, $byline = null);
 	 * @return \ILIAS\UI\Component\Input\Field\Select
 	 */
 	public function select($label, array $options, $byline = null);
+
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     A multi-select is used to allow users to pick several options from a list.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Multi Select (there are many further down, I will not mark them all to not clutter the description too much.

File: src/UI/Component/Input/Field/Factory.php [hunk 3]
```diff
@@ -348,5 +347,41 @@ public function password($label, $byline = null);
...
+	 *   composition: >
+	 *     The multi-select field will render labeled checkboxes according to given options.
+	 *   effect: >
+	 *
+	 *   rivals:
+	 *     Checkbox field: Use a checkbox field for a binary yes/no choice.
+	 *     Tag field: Use a tag input when the user is able to extend the list of given options.
+	 *     Select field: Use a select input when the user's choice is limited to one option.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: or: "... the options are mutually exclusive".

File: src/UI/Component/Input/Field/Factory.php [hunk 4]
```diff
@@ -348,5 +347,41 @@ public function password($label, $byline = null);
...
+	 *     1: >
+	 *      A multi-select input SHOULD be used when a user has to choose from a finite list of options
+	 *      which cannot be extended by the user's input and where more than one choice can be made.
+	 *     2: >
+	 *      A multi-select input MUST NOT be used whenever a user has to perform a binary choice where
+	 *      option is automatically the inverse of the other. A Checkbox MUST be used in this case.
+	 *     3: >
+	 *      A multi-select input MUST NOT be used whenever a user has to perform a choice from a list of
+	 *      options where only one option can be selected. A Select MUST be used in this case
+	 *
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: wording:
- Also add here the only postive... rule you proposed for the Radio Input

**Review by Amstutz — CHANGES_REQUESTED** [review]
For JF: Some minor details in the description. 

For the Merge: Please add the tests.

File: src/UI/Implementation/Component/Input/Field/Renderer.php [hunk 5]
```diff
@@ -353,6 +361,26 @@ public function renderSelectInput(Template $tpl, Select $input)
 		return $tpl;
 	}
 
+	public function renderMultiSelectInput(Template $tpl, MultiSelect $input) : Template	{
+		global $DIC;
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Input/Field/Renderer.php]: Why/how is this used here?

**Comment by matthiaskunkel** [comment]
Jour Fixe, 27 AUG 2018 : We highly appreciate this PR and accept it for 5.4.

**Comment by Amstutz** [comment]
@nhaagen all lights are green for merge. Resolve the conflicts run the tests one final time and then you may click the merge button to avoid new conflicts.

**Review by Amstutz — APPROVED** [review]

**Comment by nhaagen** [comment]
Thanks, @Amstutz. 
Tests are failing horribly, though.
Manually checking in examples, everything is fine...

**Comment by klees** [comment]
Hi @nhaagen, @fneumann merged the PR fixing the test failures in the meantime. Regards!

**Comment by nhaagen** [comment]
Hi @klees, @fneumann , this looks much better, now. Would you please hit the button? Thanks.

**Pull request merged**